pkg/upstream/cluster/loadbalancer.go

```diff
@@ -624,3 +626,144 @@ func (lb *reqRoundRobinLoadBalancer) IsExistsHosts(metadata api.MetadataMatchCri
+	fallback := false
+	hosts.Range(func(host types.Host) bool {
+		// Check whether `UpstreamRequestDuration` is enabled,
+		// or fallback to WRR because all hosts have same duration.
+		if _, ok := host.HostStats().UpstreamRequestDuration.(gometrics.NilHistogram); ok {
+			fallback = true
+			return false
+		}
+		return true
+	})
+
+	lb := &shortestResponseLoadBalancer{
+		hosts:           hosts,
+		rand:            rand.New(rand.NewSource(time.Now().UnixNano())),
+		choice:          default_choice,
+		wrrLoadBalancer: newWRRLoadBalancer(info, hosts),
+		fallback:        fallback,
+	}
+	return lb
+}
+
+func (lb *shortestResponseLoadBalancer) ChooseHost(context types.LoadBalancerContext) types.Host {
+	if lb.fallback {
+		return lb.wrrLoadBalancer.ChooseHost(context)
+	}
+
+	hs := lb.hosts
+	total := hs.Size()
+
+	if total == 0 {
+		return nil
+	}
+	if total == 1 {
+		host := hs.Get(0)
+		if host.Health() {
+			return host
+		}
+		return nil
+	}
+
+	var candidate types.Host
+
+	// If `total` is less than or equal to `choice`, we can iterate over all elements directly.
+	if total <= int(lb.choice) {
+		log.DefaultLogger.Debugf("[lb][shortest_response] total %d <= choice %d, using iterate", total, lb.choice)
+		candidate = lb.iterateChoose()
+	} else {
+		log.DefaultLogger.Debugf("[lb][shortest_response] total %d > choice %d, using random", total, lb.choice)
+		candidate = lb.randomChoose()
+		if candidate == nil {
+			log.DefaultLogger.Warnf("[lb][shortest_response] no host chosen after %d choice, fallback to WRR", lb.choice)
+			return lb.wrrLoadBalancer.ChooseHost(context)
+		}
+	}
+
+	return candidate
+}
+
+func (lb *shortestResponseLoadBalancer) IsExistsHosts(_ api.MetadataMatchCriteria) bool {
+	return lb.hosts.Size() > 0
+}
+
+func (lb *shortestResponseLoadBalancer) HostNum(_ api.MetadataMatchCriteria) int {
+	return lb.hosts.Size()
+}
+
+func (lb *shortestResponseLoadBalancer) iterateChoose() types.Host {
+	var candidate types.Host
+	var candidateScore float64
+
+	lb.hosts.Range(func(temp types.Host) bool {
+		if !temp.Health() {
+			return false
+		}
+
+		tempScore := shortestResponseScore(temp)
+		if candidate == nil || (tempScore < candidateScore || (tempScore == candidateScore && (temp.Weight() >= candidate.Weight()))) {
+			candidate = temp
+			candidateScore = tempScore
+		}
+
+		return true
+	})
+
+	if candidate != nil {
+		log.DefaultLogger.Debugf("[lb][shortest_response] choose host %s with score %d", candidate.AddressString(), candidateScore)
+	}
+	return candidate
+}
+
+// Choose `choice` times and return the best one.
+// See The Power of Two Random Choices: A Survey of Techniques and Results
+//
+//	http://www.eecs.harvard.edu/~michaelm/postscripts/handbook2001.pdf
+func (lb *shortestResponseLoadBalancer) randomChoose() types.Host {
+	total := lb.hosts.Size()
+
+	var candidate types.Host
+	var candidateScore float64
+
+	for i := 0; i < int(lb.choice); i++ {
+		lb.mutex.Lock()
+		idx := lb.rand.Intn(total)
+		lb.mutex.Unlock()
```

Comment:
The standard P2C requires the chosen host to be unique:
> For every request the balancer needs to route, it picks two unique instances at random.

https://blog.twitter.com/engineering/en_us/topics/infrastructure/2019/daperture-load-balancer

Maybe you are using the trace-off from the envoy? Like https://github.com/envoyproxy/envoy/issues/14859#issuecomment-770662594.

If so, let's add a comment for it.

Reply:
This is deliberate, since we only use random selection when we guarantee a minimum of 3 elements.

In fact, it is very easy to ensure a unique random selection. Select k elements from N elements. In the i-th selection, we only need to use the following method to ensure a unique random selection

$$
P(i) = Rnd(P(i - 1), N - (k - i) + 1)
$$

Reply:
> This is deliberate, since we only use random selection when we guarantee a minimum of 3 elements.

I am sorry that I can't get you.
Even if the total is larger than choices, the same element may be chosen repeatedly if we choose it randomly.

Reply:
All right, I have guaranteed that the random selection will not have duplicate elements in newest commit